File: leetcode/0946_validate_stack_sequences.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <cassert>
#include <stack>
#include <vector>

using namespace std;

class Solution {
 public:
  bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
    size_t i = 0, j = 0;
    stack<int> stack;

    while (i < pushed.size() || j < popped.size()) {
      if (j < popped.size() && !stack.empty() && stack.top() == popped[j]) {
        stack.pop();
        j++;
      } else if (i < pushed.size()) {
        stack.push(pushed[i]);
        i++;
      } else {
        return false;
      }
    }

    return i == pushed.size() && j == popped.size();
  }
};

int main() {
  auto solution = Solution();

  vector<int> pushed = {1, 2, 3, 4, 5};
  vector<int> popped = {4, 5, 3, 2, 1};

  assert(solution.validateStackSequences(pushed, popped));
}

File: leetcode/0946_validate_stack_sequences.py
# -*- coding: utf-8 -*-

from collections import deque


class Solution:
    def validateStackSequences(self, pushed, popped):
        i, j, stack = 0, 0, deque()
        while i < len(pushed) or j < len(popped):
            if j < len(popped) and stack and stack[-1] == popped[j]:
                stack.pop()
                j += 1
            elif i < len(pushed):
                stack.append(pushed[i])
                i += 1
            else:
                return False
        return i == len(pushed) and j == len(popped)


if __name__ == '__main__':
    solution = Solution()

    assert solution.validateStackSequences([1, 2, 3, 4, 5], [4, 5, 3, 2, 1])
    assert not solution.validateStackSequences([1, 2, 3, 4, 5], [4, 3, 5, 1, 2])
